Dedicated Build Pipeline

DXCanas opened the pull request:
Making this buildkite pipeline independent of Kolibri's. Makes asynchronous + conditional builds possible. 

Currently some issues with the `docker-test` command. I don't imagine it's working any better in the main Kolibri repo though. Happy to remove completely if that seems like the best move.

Upon merge, I'll likely perform some cleanup on Kolibri's pipeline so that this is no longer included.

DXCanas commented:
Was going to make this a draft, but muscle memory had me do `ctrl+enter` to create a regular PR. 

@benjaoming If you can advise on how to fix the user issues described in slack quickly, I'll happily include them here. Happy to drop `docker-test` generally, though.

benjaoming commented on build_scripts/docker_build.sh:
```diff
@@ -0,0 +1,13 @@
+
+set -euo pipefail
+
+CIDFILE=docker-deb.cid
```
Wondering if you can add some comments here about what a CID file is and why?

benjaoming commented on .buildkite/build.sh:
```diff
@@ -0,0 +1,21 @@
+  echo "--- Downloading from triggered build"
+  buildkite-agent artifact download "dist/*.tar.gz" . --build $LE_TRIGGERED_FROM_BUILD_ID
+  mv dist build_src
+  if [[ -z $BUILDKITE_TRIGGERED_FROM_BUILD_ID ]]
```
```suggestion
  # If $BUILDKITE_TRIGGERED_FROM_BUILD_ID is zero-length, no build was triggered
  if [[ -z $BUILDKITE_TRIGGERED_FROM_BUILD_ID ]]
```

DXCanas replied:
This script is pretty much only to ever run using Buildkite, and that's a buildkite functionality. That var will only ever be an empty string or a populated string, and I don't think anyone who doesn't know what they're doing should be editing this file. 

Not sure this comment is worth it?

benjaoming replied:
> I don't think anyone who doesn't know what they're doing should be editing this file

Well, it's easier to know what one is doing if the code has helpful comments :)

Doesn't matter.. if it's likely to change soon, it's not worth spending time on, I guess.

benjaoming commented on .buildkite/build.sh:
```diff
@@ -0,0 +1,21 @@
+
+make docker-deb
+
+# Upload built kolibri windows installer at buildkite artifact.
```
```suggestion
# Upload all built .deb files (one) as a buildkite artifact.
```

DXCanas replied:
nice catch

benjaoming commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+    add-apt-repository -y -u -s ppa:learningequality/kolibri-proposed
+
+RUN apt-get update && \
+    DEBIAN_FRONTEND=noninteractive apt-get -y build-dep kolibri
```
In the unlikely event that we change the build dependencies, this method would fail.

But good news! We can also install them from the `debian/control` file:

```
# Installs the build dependencies specified in debian/control
mk-build-deps --install --tool='apt-get -o Debug::pkgProblemResolver=yes --no-install-recommends --yes' debian/control
```

benjaoming replied:
This also means that we can eliminate the need to add a PPA and run `apt-get update` :)

benjaoming replied:
It was also mentioned that `apt install equivs` might be necessary: https://stackoverflow.com/a/19432425/405682

DXCanas replied:
You mind adding this in a later PR? I'm primarily porting the build scripts over from the Kolibri repo, and would like this to be in, at least as a first pass, ASAP.

benjaoming commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+    make \
+    bash \
+    tar \
+    python3
```
Is this `apk` stuff supposed to be here?

DXCanas replied:
Yep. This is running off of an Alpine image, and `apk` is Alpine's package manager. We need those packages.

benjaoming replied:
Ah okay, I got confused because of `FROM ubuntu:bionic` below where the build dependencies for the Debian installer are then installed.

DXCanas replied:
NP! And in case you're interested in further reading: https://docs.docker.com/develop/develop-images/multistage-build/

benjaoming commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+RUN make clean && make dist/kolibri_archive.tar.gz build_src/VERSION
+
+
+FROM ubuntu:bionic
```
I am yet to test this, but @jredrejo do you recall if building on non-Trusty was the reason why the .deb didn't work on Trusty?

(building the .deb on/for later versions of Ubuntu means that it won't work on Trusty)

This is not to say that we need to spend a lot of time getting this to work, but if it's desirable, it would be nice to know if it could work... gonna check back!

benjaoming replied:
Tested this - 14.04 needs to be upgraded, and it can run the version built on Buildkite, also did an extra local build of 0.13.2 on Bionic and it worked, too.

The failure to run on 14.04 is reproducible before running `apt upgrade`, I'm not sure about which version of dpkg that fixed it, but it doesn't matter. I'll file a docs issue about 14.04 support.

benjaoming commented on Makefile:
```diff
@@ -0,0 +1,48 @@
+.ONESHELL:
```
Thanks, I feel like I just learned something really really valuable about Make :)

benjaoming commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
 # This is by benjaoming's convention the directory to use when dev'ing locally
 build/
+build_src/
+dist/
```
```suggestion
# Docker working directory (shared with docker containers)
dist/
# Docker CID files
*.cid
```

DXCanas replied:
Not entirely true, except for the `debian-test` case. I can add comments, if you think they're important.

benjaoming replied:
Ah okay, then nevermind the suggestion - I find it nice to annotate `.gitignore`, it's a nice way to quickly figure out why there is some weird directory taking up space that isn't tracked in Git.

benjaoming commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
 # This is by benjaoming's convention the directory to use when dev'ing locally
 build/
+build_src/
```
```suggestion
# Working directory for Kolibri upstream sources (used by Docker)
build_src/
```

benjaoming commented:
Am curious to learn about this, what were the issues with Make `.ONESHELL` that had you switching it on and off?

DXCanas commented on build_scripts/docker_build.sh:
```diff
@@ -0,0 +1,13 @@
+
+set -euo pipefail
+
+CIDFILE=docker-deb.cid
```
There's a good amount of background info behind this specific change. I'll try not to write a novel, but it has been an ongoing project for several months.

benjaoming replied:
Thanks a lot! It's helpful to know this! :pray:

DXCanas commented on build_scripts/docker_test.sh:
```diff
@@ -0,0 +1,13 @@
+
+# set -euo pipefail
+
+DOCKER_IMAGES="ubuntu:xenial ubuntu:bionic ubuntu:trusty"
```
Sure! I can't take all the credit - I was just porting over your old `docker-test` method.

benjaoming replied:
Oh yeesh, totally forgot writing that part :)

DXCanas commented on .buildkite/build.sh:
```diff
@@ -0,0 +1,21 @@
+
+mkdir build_src
+
+if [[ $LE_TRIGGERED_FROM_BUILD_ID ]]
```
This is part of the buildkite build infrastructure I've been working on over the past couple months. Each repo has a pipeline with this defined - I'm not expecting many people to venture into the `.buildkite` directory without knowing what they're doing, and things are still in enough flux that I'd rather not document this in a comment for every repo only to have to change it later. 

The pipeline can be triggered by Kolibri's pipeline. Kolibri's pipeline sets this envvar up for every installer's pipeline trigger. The build ID Kolibri's build UID, assigned by Buildkite. Passing it along allows this pipeline to know what "parent" build triggered it and to download the appropriate WHL from that pipeline.

benjaoming replied:
Thanks for the explanation :pray: 

When putting code in different repositories, it's nice to have comments to explain where the environment originates from. Otherwise it takes more time for others to quick assess what it is and what it does. You can always briefly refer to an external documentation, it's just for the reader to have a fair chance to understand where to go next?

> Environment inherited from external Buildkite pipeline defined in: https://github.com/learningequality/blah

DXCanas replied:
I'll consider it, dude. I hear what you're saying. But like I said above: things are still in flux, and no one who doesn't know what this line is doing should be messing with it. 

Most of the team is aware that the build pipeline is being restructured into a multi-repo contraption, and makes a point to avoid touching these parts as they know it's under active dev and part of larger ecosystem.

DXCanas commented:
> Am curious to learn about this, what were the issues with Make `.ONESHELL` that had you switching it on and off?

GNU Make version discrepancies. I'm moving towards having `bash` and `docker` be the 2 deps to get a build script running for this reason. Would make it a lot easier to horizontally scale the number of build agents, too.

benjaoming commented:
> The main build machine is running a version that doesn't support it, and I imagine other build machines added down the line won't reliably have it either. 

If it helps, there might even be reason to run this with `ubuntu:trusty`, so the backwards compatibility won't be in vain. I'll get back on that! Thanks for explaining... so that's why we can `make nice-things` :)

DXCanas commented:
@jredrejo any ideas regarding the test failures in buildkite? I'm not getting them locally

benjaoming commented:
@DXCanas 

Not that I have a solution... but it seems that the auto-created version of Kolibri doesn't work in the command line entry point (which is built by setuptools). This might be necessary to escalate upstream to Kolibri, as it seems like the version itself adheres to PEP440, but somehow doesn't work, and I haven't seen this error for a long long time.

It somehow doesn't parse the "local version" part of the version in `0.13.2.dev0+git.20200506013330`.

See: https://www.python.org/dev/peps/pep-0440/#id81

```
Traceback (most recent call last):
  File "/usr/bin/kolibri", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 2749, in <module>
    working_set = WorkingSet._build_master()
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 444, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 725, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 595, in resolve
    requirements = list(requirements)[::-1]  # set up the stack
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 2605, in parse_requirements
    line, p, specs = scan_list(VERSION,LINE_END,line,p,(1,2),"version spec")
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 2583, in scan_list
    "Expected ',' or end-of-list in",line,"at",line[p:]
ValueError: ("Expected ',' or end-of-list in", 'kolibri==0.13.2.dev0+git.20200506013330', 'at', '+git.20200506013330')
make: *** [docker-test] Error 1
```

DXCanas commented:
It’s bizarre to me that this runs fine when I do the whole thing in docker locally. 

> On May 7, 2020, at 4:28 AM, Benjamin Balder Bach <notifications@github.com> wrote:
> 
> ﻿
> @DXCanas
> 
> Not that I have a solution... but it seems that the auto-created version of Kolibri doesn't work in the command line entry point (which is built by setuptools). This might be necessary to escalate upstream to Kolibri, as it seems like the version itself adheres to PEP440, but somehow doesn't work, and I haven't seen this error for a long long time.
> 
> It somehow doesn't parse the "local version" part of the version in 0.13.2.dev0+git.20200506013330.
> 
> See: https://www.python.org/dev/peps/pep-0440/#id81
> 
> Traceback (most recent call last):
>   File "/usr/bin/kolibri", line 6, in <module>
>     from pkg_resources import load_entry_point
>   File "/usr/lib/python3/dist-packages/pkg_resources.py", line 2749, in <module>
>     working_set = WorkingSet._build_master()
>   File "/usr/lib/python3/dist-packages/pkg_resources.py", line 444, in _build_master
>     ws.require(__requires__)
>   File "/usr/lib/python3/dist-packages/pkg_resources.py", line 725, in require
>     needed = self.resolve(parse_requirements(requirements))
>   File "/usr/lib/python3/dist-packages/pkg_resources.py", line 595, in resolve
>     requirements = list(requirements)[::-1]  # set up the stack
>   File "/usr/lib/python3/dist-packages/pkg_resources.py", line 2605, in parse_requirements
>     line, p, specs = scan_list(VERSION,LINE_END,line,p,(1,2),"version spec")
>   File "/usr/lib/python3/dist-packages/pkg_resources.py", line 2583, in scan_list
>     "Expected ',' or end-of-list in",line,"at",line[p:]
> ValueError: ("Expected ',' or end-of-list in", 'kolibri==0.13.2.dev0+git.20200506013330', 'at', '+git.20200506013330')
> make: *** [docker-test] Error 1
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

benjaoming commented:
@DXCanas can you access that Docker container and get the output of `kolibri --version` for comparison?

DXCanas commented:
Seeing an issue with the build step locally that I hadn't seen before because of all the logs and lack of tty. Going to try and resolve it to see if this the same problem manifesting itself differently in the different environments.

jredrejo commented on build_tools/test.sh:
```diff
@@ -0,0 +1,21 @@
+
+
+apt-get -y update
+apt-get install -y python3-pkg-resources
```
It should be
`export DEBIAN_FRONTEND=noninteractive`

jredrejo commented on build_tools/test.sh:
```diff
@@ -0,0 +1,21 @@
+
+# BOUND_DIR set in build_tools/docker_test.sh
+debconf-set-selections $BOUND_DIR/build_tools/preseed.cfg
+dpkg -i $BOUND_DIR/dist/*.deb
```
or
`DEBIAN_FRONTEND=noninteractive dpkg -i $BOUND_DIR/dist/*.deb`

DXCanas replied:
I thought that variable was set above, on line 8? But you're saying it needs to be `export`ed?

jredrejo replied:
no, either exported or used this way. It was just a suggestion if you didn't want to do the export.

benjaoming commented:
Trusty 14.04 indeed had issues with parsing complicated but PEP 440 compliant version strings! I don't remember how we've worked around that previously.. but we haven't been shipping anything with a local build string, so I think it's been given zero priority to look into.

DXCanas commented:
Alright. I've made some modifications and I think I've addressed all comments. Apart from fixes, I'd rather not add any more features to this PR :slightly_smiling_face: 

Thanks guys!

benjaoming commented:
@DXCanas can the artifacts be made public? It's really helpful for testing.

```
$ wget https://buildkite.com/organizations/learningequality/pipelines/kolibri-debian/builds/31/jobs/3c66eb54-5bc9-4bc7-8564-e63a57795c31/artifacts/997a9e44-0103-4e3b-a8f2-0f0353045598
--2020-05-10 18:48:00--  https://buildkite.com/organizations/learningequality/pipelines/kolibri-debian/builds/31/jobs/3c66eb54-5bc9-4bc7-8564-e63a57795c31/artifacts/997a9e44-0103-4e3b-a8f2-0f0353045598
Resolving buildkite.com (buildkite.com)... 34.196.162.119, 3.231.246.139
Connecting to buildkite.com (buildkite.com)|34.196.162.119|:443... connected.
HTTP request sent, awaiting response... 403 Forbidden
2020-05-10 18:48:01 ERROR 403: Forbidden.
```
